packages/stream/types.d.ts

```diff
@@ -0,0 +1,23 @@
+export interface AsyncQueue<TValue> {
+  put(value: TValue | Promise<TValue>): void,
+  get(): Promise<TValue>,
+}
+
+// Stream is nearly identical to AsyncGenerator and AsyncGenerator should
+// probably be identical to this definition of Stream.
+// Stream does not make the mistake of conflating the read and write return
+// types.
+export interface Stream<TRead, TWrite, TReadReturn, TWriteReturn> {
+  next(value: TWrite): Promise<IteratorResult<TRead, TReadReturn>>,
+  return(value: TWriteReturn): Promise<IteratorResult<TRead, TReadReturn>>,
+  throw(error: Error): Promise<IteratorResult<TRead, TReadReturn>>,
+  [Symbol.asyncIterator](): Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
+}
+
+export type Reader<TRead> = Stream<TRead, undefined, undefined, undefined>;
+export type Writer<TWrite> = Stream<undefined, TWrite, undefined, undefined>;
+
+export type makePipeFn<T, U, TReturn, UReturn> = () => [
```

Comment:
How about
```suggestion
export type MakePipeFn<T, U, TReturn, UReturn> = () => [
```
?

Reply:
I ended up needing to do this in a very different way and these types do not exist now.